README.md
# preview-scriptogram

> A simple way to preview your Scriptogram blog offline.

v0.0.2

## Dependencies

	nodejs
	npm

### NPM dependencies

    "express": "4.12.x",
    "marked": "0.3.x",
    "mustache": "2.0.x"

## Set-up

 1. Run `npm install` after download to make sure you have all dependencies.
 2. Run `setup.js` with Node to create `config.json` and `context.json` files.
 3. Ensure that the `archives` and `drafts` properties in `config.json` point to the folders containing your published and unpublished posts respectively. If you using `ahem` conventions and workflow, ensure the `ahem` property is set to `true`.
 4. Enter all the values for your blog (your name, the blog's name, etc.) into the `context.json` file.

## Usage

 1. Run `preview-scriptogram.js` with Node.
 2. Visit `127.0.0.1:1337/preview`.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and the standard outlined at [Keep a CHANGELOG](keepachangelog.com).

## [Unreleased]
### Changed
- No changes as yet.

## [0.0.2] -- 2015-05-25
### Added
 - Property in initialised `config.json` via `setup.js`. This allows users to opt into the [`ahem` workflow conventions].

## [0.0.1] -- 2015-05-25
### Added
 - Date display. Dates now rendered as they are in Scriptogram itself, based on `dates.txt` for each post.
 - Ordering of posts; previews will now show in reverse-chronological order.

### Changed
 - Title display. Instead of a test title, actual title pulled from first `h1` of main post file.
 - Tag display. Instead of a test tag, actual tags are now pulled from each post's corresponding `tags.txt` file.

### Fixed
 - Archive location. This was hard-coded in one place before; now the folder is controlled completely by `config.json`.

## 0.0.0 -- 2015-05-17
### Added
 - Basic preview script of files stored in specified `archives` location according to [`ahem` workflow conventions].
 - Ability to see JSON lists of `archives` and `drafts` files in pretty-printed plain-text.


[`ahem` workflow conventions]: https://github.com/guypursey/ahem

[0.0.2]: https://github.com/guypursey/preview-scriptogram/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/guypursey/preview-scriptogram/compare/v0.0.0...v0.0.1
[Unreleased]: https://github.com/guypursey/preview-scriptogram/compare/v0.0.1...HEAD

setup.js
// To set up `preview-scriptogram` for the first time, run this script with Node.js.

var fs = require("fs");

fs.readFile("context.json", function (err, data) {
	if (err) {
		// If file does not exist, write the file.

		// Basic values required for Scriptogram.
		var context = {
			"author": "",
			"title": "",
			"blog_title": "",
			"posts": [],
			"base_url": "",
			"css": "",
			"profile_image": "",
			"cover_image": "",
			"accent_color": "",
			"theme": "basic",
			"pages": [],
			"page": "",
			"is_archive": false,
			"pagination": ""
		};

		// Stringify context object and write to new context file.
		fs.writeFile("context.json", JSON.stringify(context, null, "\t") + "\n");

		console.log("`context.json` now created. Populate with own values to run the preview.");

	} else {
		console.log("`context.json` already exists. Please delete and run this script again to reinitialise.");

	}
});

fs.readFile("config.json", function (err, data) {
	if (err) {
		// If file does not exist, write the file.

		// Basic config options required.
		var config = {
			"archives": "../content/archives/",
			"drafts": "../content/drafts/",
			"ahem": false
		};

		// Stringify context object and write to new context file.
		fs.writeFile("config.json", JSON.stringify(config, null, "\t") + "\n");

		console.log("`config.json` now created. Please change the values to match your setup.");

	} else {
		console.log("`config.json` already exists. Please delete and run this script again to reinitialise.");

	}
});

render_with_express.js
var fs = require("fs"),
	express = require("express"),
	app = express(),
	marked = require("marked"),
	mustache = require("mustache"),
	template = fs.readFileSync("./themes/basic/main.html", { "encoding": "utf-8" }),
	archives,
	drafts,
	get_context = function () {
		var context;
		try {
			context = fs.readFileSync("context.json", { "encoding": "utf-8" });
		} catch (e) {
			console.log("ERROR", "Could not find `context.json`. Please run `setup.js`.")
		}
		return context ? JSON.parse(context) : {};
	},
	tripled_variables = ["css", "content", "profile_image"],
	context = get_context();

// By inserting a space in each triplet of braces, this prevents an error with the CSS and Mustache template.
template = template.replace(/\}\}\}/g, "}} }");

// Because Scriptogram only ever uses double-braced variables, replace these with triple-braced.
tripled_variables.forEach(function (v, i, a) {
	template = template.replace(new RegExp("\{\{(" + v + ")\}\}", "g"), "{{{$1}}}");
});

app.use("/themes", express.static('themes'));

app.use(function (req, res, next) {

	fs.readdir("../content/archives/", function (err, files) {
		if (err) {
			res.sendStatus(500); // equivalent to res.status(500).send('Internal Server Error')
		} else {
			archives = files.filter(function (e, i, a) {
				return (e.match(/^\d{12}/));
			});
			context = get_context();
			archives.forEach(function(v, i, a) {
				var current_post = {},
					post_file = fs.readFileSync("../content/archives/" + v  + "/" + v + ".md", { "encoding": "utf-8" }),
					tags_file = fs.readFileSync("../content/archives/" + v  + "/" + "tags.txt", { "encoding": "utf-8" });

				current_post["title"] = "Test post";
				current_post["content"] = marked(post_file);
				current_post["tags"] = [{ "name": "test" }];

				if (context.hasOwnProperty("posts")) {
					context.posts.push(current_post);
				}

			});

			// TODO: sort out synchronicity
			next();
		}
	});
});

app.use("/draft", function (req, res, next) {
	fs.readdir("../content/drafts/", function (err, files) {
		if (err) {
			res.sendStatus(500); // equivalent to res.status(500).send('Internal Server Error')
		} else {
			drafts = files.filter(function (e, i, a) {
				return true;
			});
			next();
		}
	});
});

app.get("/", function (req, res, next) {
	res.writeHead(200, "text/html");
	res.write(JSON.stringify(archives, null, "\t"));
	next();
});

app.use("/preview", function (req, res, next) {
	res.writeHead(200, "text/html");
	res.write(mustache.render(template, context));
	next();
});

app.get("/draft", function (req, res, next) {
	res.writeHead(200, "text/html");
	res.write(JSON.stringify(archives, null, "\t"));
	res.write(JSON.stringify(drafts, null, "\t") || " ");
	next();
});

app.use(function (req, res, next) {
	res.end();
});



app.listen(1337);

find-context-variables.js
// Run in Node.js

var fs = require("fs");

fs.readFile("./themes/basic/main.html", { "encoding": "utf-8" }, function (err, data) {
	if (err) return err;
	var context_array = data.match(/\{\{(.*?)\}\}/g),
		context_object = {};

	context_array.forEach((function (top_context) {
		var domain = [top_context];
		return function (v, i, a) {
			var property = v.replace(/\{\{(.*?)\}\}/, "$1"),
				valuetype,
				context_object = domain[0];

			// Domain shift or unshift based on block character.
			if (property.match(/^#/)) {
				// Create exemplary object within block.
				valuetype = context_object[property.substr(1)] || [{}];
				domain.unshift(valuetype[0]);
			} else if (property.match(/^\^/)) {
				domain.unshift(context_object);
			} else if (property.match(/^\//)) {
				domain.shift();
			} else {
				valuetype = "";
			}

			if (typeof valuetype !== "undefined") {
				property = property.replace(/^[#\^\/]/, "");
				if (context_object.hasOwnProperty(property)) {
					if (typeof context_object[property] !== typeof valuetype) {
						console.log("WARNING: Property `" + property + "` originally logged as " + typeof context_object[property] + " now found as " + typeof valuetype);
					}
				} else {
					context_object[property] = valuetype;
				}
			}
		}
	}(context_object)));

	fs.writeFile("context-array.json", JSON.stringify(context_object, null, "\t"), {"encoding": "utf-8"});

	console.log(context_object);
});


preview-scriptogram.js
var fs = require("fs"),
	express = require("express"),
	app = express(),
	marked = require("marked"),
	mustache = require("mustache"),
	months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"],
	archives,
	drafts,
	get_config = function () {
		var config;
		try {
			config = fs.readFileSync("config.json", { "encoding": "utf-8" });
		} catch (e) {
			console.log("ERROR", "Could not find `config.json`. Please run `setup.js`.")
		}
		return config ? JSON.parse(config) : {};
	},
	get_context = function () {
		var context;
		try {
			context = fs.readFileSync("context.json", { "encoding": "utf-8" });
		} catch (e) {
			console.log("ERROR", "Could not find `context.json`. Please run `setup.js`.")
		}
		return context ? JSON.parse(context) : {};
	},
	tripled_variables = ["css", "content", "profile_image"],
	config = get_config(),
	context = get_context(),
	template = context.theme
		? fs.readFileSync("./themes/" + context.theme + "/main.html", { "encoding": "utf-8" })
		: "";

// Alert if template has not been read.
if (!template) {
	console.log("Template could not be found. Check theme referenced in `context.json` is valid has corresponding folder and `main.html` under `themes` folder.")
}

// By inserting a space in each triplet of braces, this prevents an error with the CSS and Mustache template.
template = template.replace(/\}\}\}/g, "}} }");

// Because Scriptogram only ever uses double-braced variables, replace these with triple-braced.
tripled_variables.forEach(function (v, i, a) {
	template = template.replace(new RegExp("\{\{(" + v + ")\}\}", "g"), "{{{$1}}}");
});

app.use("/themes", express.static('themes'));

app.use(function (req, res, next) {

	var archive_location = config.archives || "";

	fs.readdir(archive_location, function (err, files) {
		if (err) {
			res.sendStatus(500); // equivalent to res.status(500).send('Internal Server Error')
		} else {
			archives = files.filter(function (e, i, a) {
				return (e.match(/^\d{12}/));
			});
			context = get_context();
			archives.forEach(function(v, i, a) {
				var current_post = {},
					post_file = "",
					tags_file = "",
					post_title = "";

				try {
					post_file = fs.readFileSync(archive_location + v + "/" + v + ".md", { "encoding": "utf-8" });
					tags_file = fs.readFileSync(archive_location + v + "/" + "tags.txt", { "encoding": "utf-8" });
					date_file = fs.readFileSync(archive_location + v + "/" + "date.txt", { "encoding": "utf-8"});
				} catch (e) {
					console.log("Could not find `" + v + ".md`. May be an issue with the `archives` property in `config.json`.")
				}

				post_title = post_file.match(/^[\s\n\r\t]*\#[\s\t]*([^\n]*)\n*/) || "";
				current_post["title"] =  post_title[1] || "";
				current_post["content"] = marked(post_title[0] ? post_file.replace(post_title[0], "") : "");

				current_post["tags"] = [];
				tags_file.split(/\n/g).forEach(function (v, i, a) {
					current_post.tags.push({ "name": v });
				});
				current_post["if_tags"] = current_post.tags.length;

				current_post["date"] = new Date(date_file.substr(0,4), date_file.substr(5,2) - 1, date_file.substr(8,2), date_file.substr(11,2) + 1, date_file.substr(14,2), 0);
				current_post["prettydate"] = months[current_post["date"].getMonth()] + " " + (+(date_file.substr(8,2))) + ", " + date_file.substr(0,4);

				if (context.hasOwnProperty("posts")) {
					context.posts.push(current_post);
				}

			});

			context.posts.sort(function (a, b) {
				return b.date - a.date;
			});

			// TODO: sort out synchronicity
			next();
		}
	});
});

app.use("/draft", function (req, res, next) {
	fs.readdir(config.drafts || "", function (err, files) {
		if (err) {
			res.sendStatus(500); // equivalent to res.status(500).send('Internal Server Error')
		} else {
			drafts = files.filter(function (e, i, a) {
				return true;
			});
			next();
		}
	});
});

app.get("/", function (req, res, next) {
	res.writeHead(200, "text/html");
	res.write(JSON.stringify(archives, null, "\t"));
	next();
});

app.use("/preview", function (req, res, next) {
	res.writeHead(200, "text/html");
	res.write(mustache.render(template, context));
	next();
});

app.get("/draft", function (req, res, next) {
	res.writeHead(200, "text/html");
	res.write(JSON.stringify(archives, null, "\t"));
	res.write(JSON.stringify(drafts, null, "\t") || " ");
	next();
});

app.use(function (req, res, next) {
	res.end();
});



app.listen(1337);